### PR TITLE
First attempt at __unique_stable_name

### PR DESCRIPTION
**Pull request by erichkeane**
This implementation uses the keyword __unique_stable_name as an operator
that can take a type or expression, and will result in a constexpr
constant character array that (hopefully) uniquely represents the type
or type-of-expression being passed to it.

The unique representation is the normal mangled name of the type, except
in the cases of a lambda, where the ordinal "number" is just replaced
with LINE->COL.  Macro expansions are also appended with '~' followed by
the LINE->COL that it appears in.

For example, a lambda declared in
'main' would look like _ZTSZ4mainEUlvE25->12 (line 25, column 12).

A Lambda that comes from a macro looks like:
_ZTSZ4mainEUlvE45->3~18->32 (macro invoked on 45/3, lambda defined
inside the macro on line 18, column 32).

Template instantiations based on the lambda would result in a name that
contains the lambda mangling, for example:
_ZTSZ3bazIZ4mainEUlvE25->12EvvEUlvE14->12

A function template named 'baz' that is instantiated with a lambda
declared in 'main' on line 25/col 12 has another macro in it, declared
on line 14, column 12.

Signed-off-by: Erich Keane <erich.keane@intel.com>

### CODE REVIEW

**Comment by bader** [comment]
+@hfinkel.
We are planning to use this built-in to make naming kernels by user optional in case both device and host compilers support this built-in function.
You feedback is very welcome.

**Comment by hfinkel** [comment]
> +@hfinkel.
> We are planning to use this built-in to make naming kernels by user optional in case both device and host compilers support this built-in function.
> You feedback is very welcome.

Thanks, @bader (and @erichkeane )! Making the kernel names optional will really help our adoption of SYCL.

Are you also working on Microsoft mangling support? As a quick note, your clang/test/CodeGenSYCL/unique-stable-name.cpp doesn't seem to test multiple nested macros.

**Comment by erichkeane** [comment]
> 
> 
> > +@hfinkel.
> > We are planning to use this built-in to make naming kernels by user optional in case both device and host compilers support this built-in function.
> > You feedback is very welcome.
> 
> Thanks, @bader (and @erichkeane )! Making the kernel names optional will really help our adoption of SYCL.
> 
> Are you also working on Microsoft mangling support? As a quick note, your clang/test/CodeGenSYCL/unique-stable-name.cpp doesn't seem to test multiple nested macros.

We don't require microsoft mangling, this is simply a "unique" and "stable" name that just so happens to be implemented in terms of the Itanium mangling structure.

I'll add that test when I ge ta shot.

**Comment by erichkeane** [comment]
@hfinkel Test Case Added.

File: clang/lib/Sema/SemaTemplateInstantiate.cpp [hunk 1]
```diff
@@ -1181,6 +1181,28 @@ TemplateInstantiator::TransformPredefinedExpr(PredefinedExpr *E) {
...
+
+    ExprResult SubExpr = getDerived().TransformExpr(E->getExpr());
+    if (SubExpr.isInvalid())
+      return ExprError();
+
+    if (!getDerived().AlwaysRebuild() && SubExpr.get() == E->getExpr())
+      return E;
+
+    return getSema().BuildUniqueStableName(E->getLocation(), SubExpr.get());
+  } else if (E->getIdentKind() == PredefinedExpr::UniqueStableNameType) {
```
Comment (bader) [inline clang/lib/Sema/SemaTemplateInstantiate.cpp]: No `else` after `return`.

File: clang/lib/AST/Expr.cpp [hunk 2]
```diff
@@ -489,6 +489,36 @@ PredefinedExpr::PredefinedExpr(SourceLocation L, QualType FNTy, IdentKind IK,
...
+PredefinedExpr::PredefinedExpr(SourceLocation L, QualType FNTy, IdentKind IK,
+                               TypeSourceInfo *Info)
+    : Expr(PredefinedExprClass, FNTy, VK_LValue, OK_Ordinary,
+           FNTy->isDependentType(), FNTy->isDependentType(),
+           FNTy->isInstantiationDependentType(),
+           /*ContainsUnexpandedParameterPack=*/false) {
+  PredefinedExprBits.Kind = IK;
+  assert((getIdentKind() == IK) &&
+         "IdentKind do not fit in PredefinedExprBitfields!");
+  assert(IK == UniqueStableNameType && "Wrong Thing!");
```
Comment (keryell) [inline clang/lib/AST/Expr.cpp]: "Wrong Kind!"?

File: clang/lib/AST/ItaniumMangle.cpp [hunk 3]
```diff
@@ -1377,7 +1381,8 @@ void CXXNameMangler::mangleUnqualifiedName(const NamedDecl *ND,
     // <lambda-sig> ::= <template-param-decl>* <parameter-type>+
     //     # Parameter types or 'v' for 'void'.
     if (const CXXRecordDecl *Record = dyn_cast<CXXRecordDecl>(TD)) {
-      if (Record->isLambda() && Record->getLambdaManglingNumber()) {
+      if (Record->isLambda() && (Record->getLambdaManglingNumber() ||
+                                 Context.isUniqueNameMangler())) {
```
Comment (keryell) [inline clang/lib/AST/ItaniumMangle.cpp]: Could be more obvious to read with `||` on this line instead.

File: clang/lib/AST/ItaniumMangle.cpp [hunk 4]
```diff
@@ -1699,6 +1704,33 @@ void CXXNameMangler::mangleTemplateParamDecl(const NamedDecl *Decl) {
...
+                                   raw_ostream &Out,
+                                   const CXXRecordDecl *Lambda) {
+  SourceLocation Loc = Lambda->getLocation();
+
+  PresumedLoc PLoc = SM.getPresumedLoc(Loc);
+  Mangler.mangleNumber(PLoc.getLine());
+  Out << "->";
+  Mangler.mangleNumber(PLoc.getColumn());
+
+  while (Loc.isMacroID()) {
```
Comment (keryell) [inline clang/lib/AST/ItaniumMangle.cpp]: While I would love more comments in this PR in general, here it is even more obvious...

File: clang/lib/Sema/SemaExpr.cpp [hunk 5]
```diff
@@ -3205,6 +3205,65 @@ ExprResult Sema::BuildPredefinedExpr(SourceLocation Loc,
...
+ExprResult Sema::BuildUniqueStableName(SourceLocation OpLoc,
+                                       TypeSourceInfo *Operand) {
+  QualType ResultTy;
+  StringLiteral *SL = nullptr;
+  if (Operand->getType()->isDependentType()) {
+    ResultTy = Context.DependentTy;
+  } else {
+    std::string Str = PredefinedExpr::ComputeName(Context,
+        PredefinedExpr::UniqueStableNameType, Operand->getType());
+    llvm::APInt Length(32, Str.length()  + 1);
```
Comment (keryell) [inline clang/lib/Sema/SemaExpr.cpp]: Spurious spacing

File: clang/lib/Sema/SemaExpr.cpp [hunk 6]
```diff
@@ -3205,6 +3205,65 @@ ExprResult Sema::BuildPredefinedExpr(SourceLocation Loc,
...
+    ResultTy = Context.adjustStringLiteralBaseType(Context.CharTy.withConst());
+    ResultTy = Context.getConstantArrayType(ResultTy, Length, ArrayType::Normal,
+                                           /*IndexTypeQuals*/ 0);
```
Comment (keryell) [inline clang/lib/Sema/SemaExpr.cpp]: Indent

File: clang/lib/Sema/SemaExpr.cpp [hunk 7]
```diff
@@ -3205,6 +3205,65 @@ ExprResult Sema::BuildPredefinedExpr(SourceLocation Loc,
 
   return PredefinedExpr::Create(Context, Loc, ResTy, IK, SL);
 }
```
Comment (keryell) [inline clang/lib/Sema/SemaExpr.cpp]: New line

File: clang/lib/Sema/SemaExpr.cpp [hunk 8]
```diff
@@ -3205,6 +3205,65 @@ ExprResult Sema::BuildPredefinedExpr(SourceLocation Loc,
...
+    ResultTy = Context.getConstantArrayType(ResultTy, Length, ArrayType::Normal,
+                                           /*IndexTypeQuals*/ 0);
+    SL = StringLiteral::Create(Context, Str, StringLiteral::Ascii,
+                               /*Pascal*/ false, ResultTy, OpLoc);
+  }
+
+  return PredefinedExpr::Create(Context, OpLoc, ResultTy,
+                                PredefinedExpr::UniqueStableNameType, SL,
+                                Operand);
+}
```
Comment (keryell) [inline clang/lib/Sema/SemaExpr.cpp]: New line

File: clang/include/clang/Basic/TokenKinds.def [hunk 9]
```diff
@@ -436,6 +436,9 @@ KEYWORD(L__FUNCSIG__                , KEYMS)
 TYPE_TRAIT_1(__is_interface_class, IsInterfaceClass, KEYMS)
 TYPE_TRAIT_1(__is_sealed, IsSealed, KEYMS)
 
+// Extensions for SYCL.
```
[the remark this answers is not in the document]
Comment (erichkeane) [reply clang/include/clang/Basic/TokenKinds.def]: This is a category heading, it isn't directly associated with the keyword itself.  If you see the rest of the file, that is the format of comments in this file.

File: clang/lib/AST/ItaniumMangle.cpp (hunk 3 above)
Comment (erichkeane) [reply clang/lib/AST/ItaniumMangle.cpp]: That is the output of Clang-Format.  For better or worse.

**Comment by erichkeane** [comment]
> 
> 
> Super useful!
> I would love more comments.
> Add some documentation so it can land to https://clang.llvm.org/docs/LanguageExtensions.html
> It would be nice to work with ISO C++ SG7 to have this feature in the standard.

Patch incoming.  

A feature like this fits very well in with the Reflection work, though I haven't been participating in SG7 very much lately.  This is implemented more like a builtin that would be used as a building block for such a proposal.

As far as putting this in LangExtensions, Clang typically doesn't document builtins in lang-extensions.  This is a building block for omitting the kernel name in SYCL functions, so I suspect this is best documented as a part of that.

The intent of this patch is to unblock that kernel naming effort, so I suspect this is a first-of-many patch for this builtin.

**Comment by hfinkel** [comment]
> > Super useful!
> > I would love more comments.
> > Add some documentation so it can land to https://clang.llvm.org/docs/LanguageExtensions.html
> > It would be nice to work with ISO C++ SG7 to have this feature in the standard.
> 
> Patch incoming.
> 
> A feature like this fits very well in with the Reflection work, though I haven't been participating in SG7 very much lately. This is implemented more like a builtin that would be used as a building block for such a proposal.
> 
> As far as putting this in LangExtensions, Clang typically doesn't document builtins in lang-extensions. This is a building block for omitting the kernel name in SYCL functions, so I suspect this is best documented as a part of that.
> 
> The intent of this patch is to unblock that kernel naming effort, so I suspect this is a first-of-many patch for this builtin.

@erichkeane , we do, see https://clang.llvm.org/docs/LanguageExtensions.html#builtin-functions - and we should document this as well.

Regarding a proposal for WG21, I'm happy to work with you and anyone else interested on a proposal. For those of you attending the next WG21 meeting, I'm also happy to chat about this in person. That seems useful.

**Comment by erichkeane** [comment]
> 
> 
> > > Super useful!
> > > I would love more comments.
> > > Add some documentation so it can land to https://clang.llvm.org/docs/LanguageExtensions.html
> > > It would be nice to work with ISO C++ SG7 to have this feature in the standard.
> > 
> > 
> > Patch incoming.
> > A feature like this fits very well in with the Reflection work, though I haven't been participating in SG7 very much lately. This is implemented more like a builtin that would be used as a building block for such a proposal.
> > As far as putting this in LangExtensions, Clang typically doesn't document builtins in lang-extensions. This is a building block for omitting the kernel name in SYCL functions, so I suspect this is best documented as a part of that.
> > The intent of this patch is to unblock that kernel naming effort, so I suspect this is a first-of-many patch for this builtin.
> 
> @erichkeane , we do, see https://clang.llvm.org/docs/LanguageExtensions.html#builtin-functions - and we should document this as well.
> 
> Regarding a proposal for WG21, I'm happy to work with you and anyone else interested on a proposal. For those of you attending the next WG21 meeting, I'm also happy to chat about this in person. That seems useful.

Huh, I've never documented one there before!  I'll work on it then.  As for the proposal, perhaps the 3 of us should discuss it in Cologne.  I'd presume that we'd need to find a way to fit this into the Reflection TS/Reflection Constexpr proposals in a way that didn't look terrible.  

Additionally, we'd likely wish to propose an actual standardized naming convention here.  This patch chooses a modified Itanium ABI, but any string would work.

File: clang/lib/AST/ItaniumMangle.cpp (hunk 3 above)
Comment (keryell) [reply clang/lib/AST/ItaniumMangle.cpp]: I see... And it does not use BreakBeforeBinaryOperators(BOS_NonAssignment)

**Comment by keryell** [comment]
I will skip the Köln meeting unfortunately... :-( Belfast?

**Comment by erichkeane** [comment]
> 
> 
> I will skip the Köln meeting unfortunately... :-( Belfast?

Cya there.

File: clang/include/clang/Basic/TokenKinds.def [hunk 10]
```diff
@@ -436,6 +436,9 @@ KEYWORD(L__FUNCSIG__                , KEYMS)
...
+KEYWORD(__unique_stable_name        , KEYSYCL)
```
Comment (rolandschulz) [inline clang/include/clang/Basic/TokenKinds.def]: It is also needed in standard C++ mode for the host compilation.

**Comment by erichkeane** [comment]
Interestingly, I got the reverse comment on the initial patch (that it was only necessary in SYCL mode).
On Sunday, June 30, 2019 rolandschulz <reply@reply.github.com> wrote:

@rolandschulz commented on this pull request.

In clang/include/clang/Basic/TokenKinds.def:
> @@ -436,6 +436,9 @@ KEYWORD(L__FUNCSIG__                , KEYMS)
 TYPE_TRAIT_1(__is_interface_class, IsInterfaceClass, KEYMS)
 TYPE_TRAIT_1(__is_sealed, IsSealed, KEYMS)
 
+// Extensions for SYCL.
+KEYWORD(__unique_stable_name        , KEYSYCL)

It is also needed in standard C++ mode for the host compilation.

File: clang/docs/LanguageExtensions.rst [hunk 11]
```diff
@@ -1950,6 +1950,30 @@ form of ``__builtin_operator_delete`` is currently available.
 These builtins are intended for use in the implementation of ``std::allocator``
 and other similar allocation libraries, and are only available in C++.
 
+``__unique_stable_name``
+------------------------
+
+``__unique_stable_name()`` is a SYCL builtin that takes a type or expression and
```
Comment (bader) [inline clang/docs/LanguageExtensions.rst]:
```suggestion
``__unique_stable_name()`` is a builtin that takes a type or expression and
```
Recent change enabled this built-in for all "languages".

File: clang/include/clang/AST/Expr.h [hunk 12]
```diff
@@ -1843,20 +1875,47 @@ class PredefinedExpr final
...
+
+  Expr *getExpr() {
+    assert(!hasFunctionName() && getIdentKind() == UniqueStableNameExpr &&
+           "Expr only valid for UniqueStableName of an Expression.");
+    return getTrailingObjects<PredefExprStorage>()->E;
+  }
+
+  const Expr *getExpr() const {
+    assert(!hasFunctionName() && getIdentKind() == UniqueStableNameExpr &&
+           "TypeSourceInfo only valid for UniqueStableName of a Type");
```
Comment (bader) [inline clang/include/clang/AST/Expr.h]:
```suggestion
           "Expr only valid for UniqueStableName of an Expression.");
```

File: clang/include/clang/Basic/TokenKinds.def (hunk 9 above)
[the remark this answers is not in the document]
Comment (bader) [reply clang/include/clang/Basic/TokenKinds.def]: Probably, we can move this built-in to some "generic" clang built-ins category.

File: clang/lib/Sema/SemaExpr.cpp (hunk 7 above)
Comment (bader) [reply clang/lib/Sema/SemaExpr.cpp]: +1

**Comment by erichkeane** [comment]
I'm not authorized to merge this pull request, so I don't know should be doing it.

**Comment by bader** [comment]
I'd like to wait a bit to see if other reviewers have additional comments to the latest version of the patch and if there are no objections, I'll merge this patch tomorrow.
@erichkeane, thanks for working on this!

**Pull request merged**